tools/clean.sh

```diff
@@ -0,0 +1,109 @@
+#!/bin/sh
+# Shell script equivalent of Metamath CLEAN tool
+# Mario Carneiro 5-Jan-2022
+
+usage() {
+  cat << HELP
+CLEAN - Trim spaces and tabs on each line in a file; convert characters
+
+This command processes spaces and tabs in each line of a file
+according to the following subcommands:
+  D - Delete all spaces and tabs
+  B - Delete spaces and tabs at the beginning of each line
+  E - Delete spaces and tabs at the end of each line
+  R - Reduce multiple spaces and tabs to one space
+  Q - Do not alter characters in quotes (ignored by T and U)
+  T - (Tab) Convert spaces to equivalent tabs (unsupported)
+  U - (Untab) Convert tabs to equivalent spaces (unsupported)
+Some other subcommands are also available:
+  P - Trim parity (8th) bit from each character
+  G - Discard garbage characters CR,FF,ESC,BS
+  C - Convert to upper case
+  L - Convert to lower case
+  V - Convert VT220 screen print frame graphics to -,|,+ characters
+Subcommands are separated by spaces, e.g., B E R Q
+Syntax:  clean.sh subcmd subcmd ... < INFILE > OUTFILE
+HELP
+}
+
+if [ $# -eq 1 ] && [ "$1" = "-h" ]; then usage; exit; fi
+if [ $# -eq 0 ]; then usage; exit 1; fi
+
+for var in "$@"; do
+  case $var in
+    D | d) allDiscard=1;;
+    B | b) leadDiscard=1;;
+    E | e) trailDiscard=1;;
+    R | r) reduce=1;;
+    Q | q) quote=1;;
+    T | t) echo "T (tab) flag unsupported"; exit 2;;
+    U | u) echo "U (untab) flag unsupported"; exit 2;;
+    P | p) trim=1;;
```

Comment:
I had to get back here to lookup what trim means. Suggest to rename it to clearParity